presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java

```diff
@@ -0,0 +1,458 @@
+ * SELECT "$data_sequence_number", * FROM table
+ * LEFT JOIN "table$equality_deletes1" d1 ON left.pk = d1.pk AND left."$data_sequence_number" < d1."$data_sequence_number" -- Find deletes by schema 1
+ * LEFT JOIN "table$equality_deletes2" d2 ON left.orderid = d1.orderid AND left."$data_sequence_number" < d2."$data_sequence_number" -- Find deletes by schema 2
+ * WHERE COALESCE(d1."$data_sequence_number", d2."data_sequence_number") IS NULL -- None of the delete files had a delete for this row
+ * </pre>
+ * Note that table$equality_deletes1 and table$equality_deletes2 are different tables, each containing only the delete files with the schema for this join.</p>
+ */
+
+public class IcebergEqualityDeleteAsJoin
+        implements ConnectorPlanOptimizer
+{
+    private final StandardFunctionResolution functionResolution;
+    private final IcebergTransactionManager transactionManager;
+    private final TypeManager typeManager;
+
+    IcebergEqualityDeleteAsJoin(StandardFunctionResolution functionResolution,
+            IcebergTransactionManager transactionManager,
+            TypeManager typeManager)
+    {
+        this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (!isDeleteToJoinPushdownEnabled(session)) {
+            return maxSubplan;
+        }
+        return rewriteWith(new DeleteAsJoinRewriter(functionResolution,
+                transactionManager, idAllocator, session, typeManager, variableAllocator), maxSubplan);
+    }
+
+    private static class DeleteAsJoinRewriter
+            extends ConnectorPlanRewriter<Void>
+    {
+        private final ConnectorSession session;
+        private final StandardFunctionResolution functionResolution;
+        private final PlanNodeIdAllocator idAllocator;
+        private final IcebergTransactionManager transactionManager;
+        private final TypeManager typeManager;
+        private final VariableAllocator variableAllocator;
+
+        public DeleteAsJoinRewriter(
+                StandardFunctionResolution functionResolution,
+                IcebergTransactionManager transactionManager,
+                PlanNodeIdAllocator idAllocator,
+                ConnectorSession session,
+                TypeManager typeManager,
+                VariableAllocator variableAllocator)
+        {
+            this.functionResolution = functionResolution;
+            this.transactionManager = transactionManager;
+            this.idAllocator = idAllocator;
+            this.session = session;
+            this.typeManager = typeManager;
+            this.variableAllocator = variableAllocator;
```

Comment:
Please validate inputs e.g. 
```suggestion
            this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
```